src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -107,33 +117,54 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& transformation)
         // Calculate cost due to {data point, target neighbors, impostors}
         // triplets.
         double eval = 0;
-        if ((iteration - 1 % range == 0) && (iteration - 1 != 0))
+        if (transformationOld.n_elem != 0)
         {
-          eval = metric.Evaluate(transformedDataset.col(i),
-                    transformedDataset.col(targetNeighbors(j, i))) -
-                 distance(l, i);
+          eval = evalOld[i] + transformationDiff * (norm(targetNeighbors(j, i)) +
+              norm(impostors(l, i)) + 2 * norm(i));
+          if (eval > -1)
+          {
+            // Calculate exact eval.
+            if (iteration - 1 % range == 0)
+            {
+              eval = metric.Evaluate(transformedDataset.col(i),
+                         transformedDataset.col(targetNeighbors(j, i))) -
+                     distance(l, i);
+            }
+            else
+            {
+              eval = metric.Evaluate(transformedDataset.col(i),
+                       transformedDataset.col(targetNeighbors(j, i))) -
+                     metric.Evaluate(transformedDataset.col(i),
+                       transformedDataset.col(impostors(l, i)));
+            }
+          }
         }
         else
         {
           eval = metric.Evaluate(transformedDataset.col(i),
-                    transformedDataset.col(targetNeighbors(j, i))) -
-                 metric.Evaluate(transformedDataset.col(i),
-                    transformedDataset.col(impostors(l, i)));
+                     transformedDataset.col(targetNeighbors(j, i))) -
+                 distance(l, i);
         }
 
         // Check bounding condition.
-        if (eval < -1)
+        if (eval <= -1)
         {
           // update bound.
           bp = l;
           break;
         }
 
+        // Update cache eval value.
+        evalOld[i] = eval;
```

Comment:
I always had my doubts regarding this, and since I was always neglecting and considering the marginal improvement as system noise. After some more confusing thoughts, I end up putting it here.

So basically this should be done at least before the `break` statement, right?